File: nodejs/selenium-file-download/scrape.js
#!/usr/bin/env node
const { Builder, Browser, By, Condition } = require('selenium-webdriver');
const fs = require('fs/promises');
const {
    AUTH = 'USER:PASS',
    TARGET_URL = 'https://calmcode.io/datasets/bigmac',
    SELECTOR = 'button.border',
    FILENAME = './testfile.csv',
    CHUNK_SIZE = '1048576',
} = process.env;

async function scrape(url = TARGET_URL, selector = SELECTOR, filename = FILENAME) {
    if (AUTH == 'USER:PASS') {
        throw new Error(`Provide Scraping Browsers credentials in AUTH`
            + ` environment variable or update the script.`);
    }
    console.log(`Connecting to Browser...`);
    const server = `https://${AUTH}@brd.superproxy.io:9515`;
    const driver = await new Builder()
        .forBrowser(Browser.CHROME)
        .usingServer(server)
        .build();
    const cdp = (name, params = {}) => driver.sendAndGetDevToolsCommand(name, params);
    try {
        console.log(`Connected! Enable file download...`);
        await cdp('Download.enable', { allowedContentTypes: ['application/octet-stream'] });
        console.log(`Navigating to ${url}...`);
        await driver.get(url);
        console.log(`Navigated! Initiating download...`);
        const initiator = await driver.findElement(By.css(selector));
        await initiator.click();
        const id = await driver.wait(new Condition('Waiting download completed', async () => {
            const last = await cdp('Download.getLastCompleted');
            return last.id;
        }));
        console.log(`Download completed! Saving it to ${filename}...`);
        const body = await getDownloadedBody(cdp, id);
        const file = await fs.open(filename, 'w');
        await file.write(body);
        await file.close();
        console.log(`Download saved! Size: ${body.length}.`);
    } finally {
        console.log(`Closing session.`);
        await driver.quit();
    }
}

async function getDownloadedBody(cdp, id, chunkSize = +CHUNK_SIZE) {
    let offset = 0;
    const parts = [];
    for (let i = 1;; i++) {
        const { body, base64Encoded, eof } =
            await cdp('Download.getDownloadedBody', { id, offset, size: chunkSize });
        const chunk = Buffer.from(body, base64Encoded ? 'base64' : 'utf8');
        console.log(`Download chunk #${i}! Size: ${chunk.length}`);
        parts.push(chunk);
        offset += chunk.length;
        if (eof) {
            return Buffer.concat(parts);
        }
    }
}

if (require.main == module) {
    scrape().catch(error => {
        console.error(error.stack || error.message || error);
        process.exit(1);
    });
}

File: python/selenium-file-download/scrape.py
#!/usr/bin/env python3
from os import environ
from base64 import standard_b64decode
from selenium.webdriver import Remote, ChromeOptions as Options
from selenium.webdriver.chromium.remote_connection import \
    ChromiumRemoteConnection as Connection
from selenium.webdriver.common.by import By
from selenium.webdriver.support.wait import WebDriverWait

AUTH = environ.get('AUTH', default='USER:PASS')
TARGET_URL = environ.get('TARGET_URL',
                         default='https://calmcode.io/datasets/bigmac')
SELECTOR = environ.get('SELECTOR', default='button.border')
FILENAME = environ.get('FILENAME', default='./testfile.csv')
CHUNK_SIZE = int(environ.get('CHUNK_SIZE', default='1048576'))


def scrape(url=TARGET_URL, selector=SELECTOR, filename=FILENAME):
    if AUTH == 'USER:PASS':
        raise Exception('Provide Scraping Browsers credentials in AUTH '
                        'environment variable or update the script.')
    print('Connecting to Browser...')
    server_addr = f'https://{AUTH}@brd.superproxy.io:9515'
    connection = Connection(server_addr, 'goog', 'chrome')
    driver = Remote(connection, options=Options())

    def cdp(cmd, params={}):
        return driver.execute('executeCdpCommand', {
            'cmd': cmd,
            'params': params,
        })['value']

    try:
        print('Connected! Enable file download...')
        cdp('Download.enable', {'allowedContentTypes': ['application/octet-stream']})
        print(f'Navigating to {url}...')
        driver.get(url)
        print('Navigated! Initiating download...')
        driver.find_element(By.CSS_SELECTOR, selector).click()
        print('Waiting for download...')
        wait = WebDriverWait(driver, timeout=60, ignored_exceptions=(KeyError))
        download_id = wait.until(lambda _: cdp('Download.getLastCompleted')['id'])
        print(f'Download completed! Saving it to {filename}')
        file_size = 0
        chunk_index = 1
        offset = 0
        with open(filename, 'wb') as file:
            while True:
                response = cdp('Download.getDownloadedBody', {
                    'id': download_id,
                    'offset': offset,
                    'size': CHUNK_SIZE,
                })
                if response['base64Encoded']:
                    chunk = standard_b64decode(response['body'])
                else:
                    chunk = bytes(response['body'], 'utf8')
                print(f'Download chunk #{chunk_index}! Size: {len(chunk)}')
                file.write(chunk)
                file_size += len(chunk)
                offset += len(chunk)
                chunk_index += 1
                if response['eof']:
                    break
        print(f'Download saved! Size: {file_size}')
    finally:
        print('Closing session.')
        driver.quit()


if __name__ == '__main__':
    scrape()
